salt/_logging/impl.py

```diff
@@ -400,51 +512,562 @@ def makeRecord(
+        )
+    if opts.get("configure_file_logger", True) and not is_logfile_handler_configured():
+        log_file_level = opts["log_level_logfile"] or opts["log_level"]
+        if log_file_level != "quiet":
+            log_file_key = opts.get("log_file_key") or "log_file"
+            setup_logfile_handler(
+                log_path=opts[log_file_key],
+                log_level=log_file_level,
+                log_format=opts["log_fmt_logfile"],
+                date_format=opts["log_datefmt_logfile"],
+                max_bytes=opts["log_rotate_max_bytes"],
+                backup_count=opts["log_rotate_backup_count"],
+                user=opts["user"],
+            )
+        else:
+            setup_logfile_handler.__handler__ = logging.NullHandler()
+    if (
+        opts.get("configure_ext_handlers", True)
+        and not is_extended_logging_configured()
+    ):
+        setup_extended_logging(opts)
+    if opts.get("configure_granular_levels", True):
+        setup_log_granular_levels(opts["log_granular_levels"])
+
+
+def shutdown_logging():
+    if is_extended_logging_configured():
+        shutdown_extended_logging()
+    if is_logfile_handler_configured():
+        shutdown_logfile_handler()
+    if is_console_handler_configured():
+        shutdown_console_handler()
+
+
+def in_mainprocess():
+    """
+    Check to see if this is the main process
+    """
+    try:
+        return in_mainprocess.__pid__ == os.getpid()
+    except AttributeError:
+        if multiprocessing.current_process().name == "MainProcess":
+            in_mainprocess.__pid__ = os.getpid()
+            return True
+        return False
+
+
+def get_lowest_log_level():
+    """
+    Get the lowest log level
+    """
+    try:
+        return set_lowest_log_level.__log_level__
+    except AttributeError:
+        return
+
+
+def set_lowest_log_level(log_level):
+    """
+    Set the lowest log level
+    """
+    set_lowest_log_level.__log_level__ = log_level
+    # Additionally set the root logger to the same level.
+    # Nothing below this level should be processed by python's logging machinery
+    logging.root.setLevel(log_level)
+
+
+def set_lowest_log_level_by_opts(opts):
+    """
+    Set the lowest log level by the passed config
+    """
+    log_levels = [
+        get_logging_level_from_string(opts.get("log_level")),
+        get_logging_level_from_string(opts.get("log_level_logfile")),
+    ]
+    for log_level in opts.get("log_granular_levels", {}).values():
+        log_levels.append(get_logging_level_from_string(log_level))
+
+    log_level = min(log_levels)
+    set_lowest_log_level(log_level)
+
+
+def __global_logging_exception_handler(
+    exc_type,
+    exc_value,
+    exc_traceback,
+    _logger=logging.getLogger(__name__),
+    _stderr=sys.__stderr__,
+    _format_exception=traceback.format_exception,
+):
+    """
+    This function will log all un-handled python exceptions.
+    """
+    if exc_type.__name__ == "KeyboardInterrupt":
```

Comment:
**question** why not `isinstance`?

Reply:
Because you'd need a reference to `KeyboardInterrupt` and if I recall correctly during interpreter shutdown, that reference gets cleaned too early.